hazelcast/src/main/java/com/hazelcast/map/impl/MapEntries.java

```diff
@@ -85,4 +77,28 @@ public int getFactoryId() {
     public int getId() {
         return MapDataSerializerHook.ENTRY_SET;
     }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        int size = entries.size();
+        out.writeInt(size);
+
+        if (size > 0) {
+            for (Map.Entry<Data, Data> o : entries) {
+                out.writeData(o.getKey());
+                out.writeData(o.getValue());
+            }
+        }
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        int size = in.readInt();
+        for (int i = 0; i < size; i++) {
```

Comment:
a small optimization: when `size > X` then you could re-create the array with right capacity - to avoid copying when growing the array-list.